plugins/languages/treesitter/treesitter.nix

```diff
@@ -6,210 +6,316 @@
+  # NOTE: We call require('nvim-treesitter.configs').setup in extraConfigLua.
+  callSetup = false;
+  # NOTE: We dont want to try installing the provided package because we install it in extraPlugins
+  installPackage = false;
 
-          custom_captures = if (cfg.customCaptures != { }) then cfg.customCaptures else null;
-        };
+  extraConfig = cfg: {
+    assertions = [
+      {
+        assertion = ((cfg.grammarPackages != [ ]) || cfg.grammarPackages == "all") -> cfg.package != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have enabled `grammarPackages`, but `plugins.treesitter.package` is `null`.
+          This package is required to choose which grammars to install.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.gccPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.gccPackage` is `null`.
+          This package is required to build the grammars at runtime.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.nodejsPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.nodejsPackage` is `null`.
+          This package is required to build the grammars at runtime.
+        '';
+      }
+      {
+        assertion = !cfg.nixGrammars -> cfg.treesitterPackage != null;
+        message = ''
+          Nixvim (plugins.treesitter): You have disabled `nixGrammars`, but `plugins.treesitter.treesitterPackage` is `null`.
+          This package is required to build the grammars at runtime.
+        '';
+      }
+    ];
 
-        incremental_selection =
-          if cfg.incrementalSelection.enable then
-            {
-              enable = true;
-              keymaps = {
-                init_selection = cfg.incrementalSelection.keymaps.initSelection;
-                node_incremental = cfg.incrementalSelection.keymaps.nodeIncremental;
-                scope_incremental = cfg.incrementalSelection.keymaps.scopeIncremental;
-                node_decremental = cfg.incrementalSelection.keymaps.nodeDecremental;
-              };
-            }
-          else
-            null;
-
-        indent = if cfg.indent then { enable = true; } else null;
-
-        ensure_installed = if cfg.nixGrammars then [ ] else cfg.ensureInstalled;
-        ignore_install = cfg.ignoreInstall;
-        parser_install_dir = cfg.parserInstallDir;
-      } // cfg.moduleConfig;
-    in
-    mkIf cfg.enable {
-      extraConfigLua =
-        (optionalString (cfg.parserInstallDir != null) ''
-          vim.opt.runtimepath:append("${cfg.parserInstallDir}")
+    extraConfigLua =
+      ''
+        do
+      ''
+      +
+        # NOTE: It MUST be at the beginning of runtimepath. Otherwise the parsers from Neovim itself
+        # is loaded that may not be compatible with the queries from the 'nvim-treesitter' plugin.
```

Comment:
```suggestion
        # NOTE: Upstream state that the parser MUST be at the beginning of runtimepath.
        # Otherwise the parsers from Neovim takes precedent, which may be incompatible with some queries.
```